keps/sig-node/3857-rro-mounts/README.md

```diff
@@ -0,0 +1,1080 @@
+ type VolumeMount struct {
+ 	// Required: This must match the Name of a Volume [above].
+ 	Name string
+-	// Optional: Defaults to false (read-write).
++	// Optional: Defaults to false (read-write), unless RecursivelyReadOnly is set to a non-nil value.
+ 	// +optional
+-	ReadOnly bool
++	ReadOnly *bool
++	// RecursivelyReadOnly specifies recursively-readonly mode.
++	// When RecursivelyReadOnly is set to a non-nil value, ReadOnly is validated to be true.
++	// When RecursivelyReadOnly is nil, ReadOnly is not validated to be true, and if ReadOnly is true,
++	// it will be a non-recursive readonly mount.
++	// +featureGate=RecursivelyReadOnlyMounts
++	// +optional
++	RecursivelyReadOnly *RecursivelyReadOnlyMode
+ 	// Required. If the path is not an absolute path (e.g. some/path) it
+ 	// will be prepended with the appropriate root prefix for the operating
+ 	// system.  On Linux this is '/', on Windows this is 'C:\'.
+@@ -1905,6 +1912,18 @@ const (
+ 	MountPropagationBidirectional MountPropagationMode = "Bidirectional"
+ )
+ 
++// RecursivelyReadOnlyMode describes recursively-readonly mode.
++type RecursivelyReadOnlyMode string
++
++const (
++	// RecursivelyReadOnlyDisabled disables recursively-readonly mode.
++	RecursivelyReadOnlyDisabled RecursivelyReadOnlyMode = "Disabled"
++	// RecursivelyReadOnlyIfPossible enables recursively-readonly mode if possible.
++	RecursivelyReadOnlyIfPossible RecursivelyReadOnlyMode = "IfPossible"
++	// RecursivelyReadOnlyEnabled enables recursively-readonly mode, or raise an error.
++	RecursivelyReadOnlyEnabled RecursivelyReadOnlyMode = "Enabled"
++)
++
+ // VolumeDevice describes a mapping of a raw block device within a container.
+ type VolumeDevice struct {
+ 	// name must match the name of a persistentVolumeClaim in the pod
+@@ -2505,6 +2524,10 @@ type ContainerStatus struct {
+ 	// +featureGate=InPlacePodVerticalScaling
+ 	// +optional
+ 	Resources *ResourceRequirements
++	// Status of volume mounts.
++	// +listType=atomic
++	// +optional
++	VolumeMounts []VolumeMountStatus
+ }
+ 
+ // PodPhase is a label for the condition of a pod at the current time.
+@@ -2584,6 +2607,24 @@ const (
+ 	PodResizeStatusInfeasible PodResizeStatus = "Infeasible"
+ )
+ 
++// VolumeMountStatus shows status of volume mounts.
++type VolumeMountStatus struct {
++	// Corresponds to the name of the original VolumeMount.
++	Name string
++	// Corresponds to the mount path of the original VolumeMount.
++	MountPath string
++	// ReadOnly must be set to true or false.
++	// A nil value in the original VolumeMount must be translated to true or false,
++	// depending on the mount result.
++	ReadOnly bool
++	// RecursivelyReadOnly must be set to Disabled, Enabled, or nil (for non-readonly mounts).
++	// An IfPossible value in the original VolumeMount must be translated to Disabled or Enabled,
++	// depending on the mount result.
++	// +featureGate=RecursivelyReadOnlyMounts
++	// +optional
++	RecursivelyReadOnly *RecursivelyReadOnlyMode
```

Comment:
What do you think about making this field more generalizable to any future read only extensions?

So instead of naming it `RecusirvelyReadOnly`, what about something like `ReadOnlyOptions` that can take a list of values like `RecusiveIfPossible`,  `RecursiveAlways`?

cc @thockin

Reply:
> So instead of naming it `RecusirvelyReadOnly`, what about something like `ReadOnlyOptions` that can take a list of values like `RecusiveIfPossible`, `RecursiveAlways`?

It is harder to validate

Reply:
I am on the fence.  We know that our ability to predice future is bad (which is why it should not have been a bool in the first place, harumph!) but I am not sure a list of options is better.  The problem is that we have to support "require" and "ifpossible" modes, and those are not orthogonal.  So now we end up with a list of options, some of which are mutually exclusive with others, which...feels like discrete fields.

I was initiallly excited by this idea, but as I wrote it out more, I liked it less.

Reply:
What should we do? Shall we just drop "IfPossible" and make it a plain bool?
(`true` will refuse to run the container when the kernel is too old)